rustworkx-core/src/token_swapper.rs

```diff
@@ -0,0 +1,556 @@
+// Licensed under the Apache License, Version 2.0 (the "License"); you may
+// not use this file except in compliance with the License. You may obtain
+// a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+// WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+// License for the specific language governing permissions and limitations
+// under the License.
+
+use rand::distributions::Uniform;
+use rand::prelude::*;
+use rand_pcg::Pcg64;
+use std::fmt::Debug;
+use std::hash::Hash;
+use std::usize::MAX;
+
+use hashbrown::HashMap;
+use petgraph::stable_graph::{NodeIndex, StableGraph};
+use petgraph::visit::{
+    EdgeCount, GraphBase, IntoEdges, IntoNeighborsDirected, IntoNodeIdentifiers, NodeCount,
+    NodeIndexable, Visitable,
+};
+use petgraph::Directed;
+use petgraph::Direction::{Incoming, Outgoing};
+
+use crate::connectivity::find_cycle;
+use crate::dictmap::*;
+use crate::shortest_path::dijkstra;
+use crate::traversal::dfs_edges;
+
+type Swap = (NodeIndex, NodeIndex);
+type Edge = (NodeIndex, NodeIndex);
+
+struct TokenSwapper<G: GraphBase>
```

Comment:
That's fair. I was just wondering because everything seemed to involve the single struct. Either way I'm fine with this structure, if a need comes up to make it public it's easy to add later